File: src/services/code-index/__tests__/fixtures/typescript/classes.ts
/**
 * Test fixture: TypeScript classes with inheritance, access modifiers, and methods
 * 
 * Tests:
 * - Class definitions
 * - Inheritance hierarchies
 * - Access modifiers (public, private, protected)
 * - Static methods and properties
 * - Abstract classes
 * - Method overriding
 */

// Base abstract class
export abstract class Entity {
  protected id: string
  protected createdAt: Date
  protected updatedAt: Date

  constructor(id: string) {
    this.id = id
    this.createdAt = new Date()
    this.updatedAt = new Date()
  }

  abstract validate(): boolean

  public getId(): string {
    return this.id
  }

  protected touch(): void {
    this.updatedAt = new Date()
  }
}

// User class extending Entity
export class User extends Entity {
  private email: string
  private passwordHash: string
  public username: string
  private roles: string[]

  constructor(id: string, email: string, username: string) {
    super(id)
    this.email = email
    this.username = username
    this.passwordHash = ''
    this.roles = ['user']
  }

  public validate(): boolean {
    return this.email.includes('@') && this.username.length > 0
  }

  public setPassword(password: string): void {
    // Simplified password hashing
    this.passwordHash = Buffer.from(password).toString('base64')
    this.touch()
  }

  public checkPassword(password: string): boolean {
    const hash = Buffer.from(password).toString('base64')
    return this.passwordHash === hash
  }

  public addRole(role: string): void {
    if (!this.roles.includes(role)) {
      this.roles.push(role)
      this.touch()
    }
  }

  public hasRole(role: string): boolean {
    return this.roles.includes(role)
  }

  public getEmail(): string {
    return this.email
  }

  private sanitizeEmail(email: string): string {
    return email.toLowerCase().trim()
  }
}

// Admin class extending User
export class AdminUser extends User {
  private permissions: Set<string>
  private static adminCount: number = 0

  constructor(id: string, email: string, username: string) {
    super(id, email, username)
    this.permissions = new Set(['read', 'write', 'delete'])
    this.addRole('admin')
    AdminUser.adminCount++
  }

  public static getAdminCount(): number {
    return AdminUser.adminCount
  }

  public grantPermission(permission: string): void {
    this.permissions.add(permission)
    this.touch()
  }

  public revokePermission(permission: string): void {
    this.permissions.delete(permission)
    this.touch()
  }

  public hasPermission(permission: string): boolean {
    return this.permissions.has(permission)
  }

  public override validate(): boolean {
    return super.validate() && this.permissions.size > 0
  }
}

// Service class with dependency injection
export class UserService {
  private users: Map<string, User>
  private static instance: UserService | null = null

  private constructor() {
    this.users = new Map()
  }

  public static getInstance(): UserService {
    if (!UserService.instance) {
      UserService.instance = new UserService()
    }
    return UserService.instance
  }

  public createUser(email: string, username: string): User {
    const id = crypto.randomUUID()
    const user = new User(id, email, username)
    this.users.set(id, user)
    return user
  }

  public getUser(id: string): User | undefined {
    return this.users.get(id)
  }

  public deleteUser(id: string): boolean {
    return this.users.delete(id)
  }

  public findByEmail(email: string): User | undefined {
    return Array.from(this.users.values()).find(u => u.getEmail() === email)
  }
}

File: src/services/code-index/__tests__/fixtures/python/classes.py
"""
Test fixture: Python classes with inheritance and magic methods

Tests:
- Class definitions
- Inheritance
- Magic methods (__init__, __str__, __repr__, etc.)
- Class methods and static methods
- Properties and decorators
- Abstract base classes
"""

from abc import ABC, abstractmethod
from typing import List, Optional
from datetime import datetime


class Entity(ABC):
    """Abstract base class for all entities"""
    
    def __init__(self, entity_id: str):
        self.id = entity_id
        self.created_at = datetime.now()
        self.updated_at = datetime.now()
    
    @abstractmethod
    def validate(self) -> bool:
        """Validate the entity"""
        pass
    
    def touch(self) -> None:
        """Update the updated_at timestamp"""
        self.updated_at = datetime.now()
    
    def __repr__(self) -> str:
        return f"{self.__class__.__name__}(id={self.id})"


class User(Entity):
    """User class with authentication"""
    
    def __init__(self, user_id: str, email: str, username: str):
        super().__init__(user_id)
        self._email = email
        self.username = username
        self._password_hash = None
        self._roles = ['user']
    
    @property
    def email(self) -> str:
        """Get user email"""
        return self._email
    
    @email.setter
    def email(self, value: str) -> None:
        """Set user email with validation"""
        if '@' not in value:
            raise ValueError("Invalid email format")
        self._email = value.lower().strip()
        self.touch()
    
    def validate(self) -> bool:
        """Validate user data"""
        return '@' in self._email and len(self.username) > 0
    
    def set_password(self, password: str) -> None:
        """Set user password (simplified hashing)"""
        import hashlib
        self._password_hash = hashlib.sha256(password.encode()).hexdigest()
        self.touch()
    
    def check_password(self, password: str) -> bool:
        """Check if password matches"""
        import hashlib
        password_hash = hashlib.sha256(password.encode()).hexdigest()
        return self._password_hash == password_hash
    
    def add_role(self, role: str) -> None:
        """Add a role to the user"""
        if role not in self._roles:
            self._roles.append(role)
            self.touch()
    
    def has_role(self, role: str) -> bool:
        """Check if user has a specific role"""
        return role in self._roles
    
    def __str__(self) -> str:
        return f"User({self.username}, {self._email})"
    
    def __eq__(self, other) -> bool:
        if not isinstance(other, User):
            return False
        return self.id == other.id


class AdminUser(User):
    """Admin user with additional permissions"""
    
    _admin_count = 0
    
    def __init__(self, user_id: str, email: str, username: str):
        super().__init__(user_id, email, username)
        self._permissions = {'read', 'write', 'delete'}
        self.add_role('admin')
        AdminUser._admin_count += 1
    
    @classmethod
    def get_admin_count(cls) -> int:
        """Get total number of admin users"""
        return cls._admin_count
    
    @staticmethod
    def is_valid_permission(permission: str) -> bool:
        """Check if permission is valid"""
        valid_permissions = {'read', 'write', 'delete', 'admin'}
        return permission in valid_permissions
    
    def grant_permission(self, permission: str) -> None:
        """Grant a permission to the admin"""
        if self.is_valid_permission(permission):
            self._permissions.add(permission)
            self.touch()
    
    def revoke_permission(self, permission: str) -> None:
        """Revoke a permission from the admin"""
        self._permissions.discard(permission)
        self.touch()
    
    def has_permission(self, permission: str) -> bool:
        """Check if admin has a specific permission"""
        return permission in self._permissions
    
    def validate(self) -> bool:
        """Validate admin user"""
        return super().validate() and len(self._permissions) > 0


class UserService:
    """Singleton service for managing users"""
    
    _instance: Optional['UserService'] = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._users = {}
        return cls._instance
    
    def create_user(self, email: str, username: str) -> User:
        """Create a new user"""
        import uuid
        user_id = str(uuid.uuid4())
        user = User(user_id, email, username)
        self._users[user_id] = user
        return user
    
    def get_user(self, user_id: str) -> Optional[User]:
        """Get user by ID"""
        return self._users.get(user_id)
    
    def delete_user(self, user_id: str) -> bool:
        """Delete a user"""
        if user_id in self._users:
            del self._users[user_id]
            return True
        return False
    
    def find_by_email(self, email: str) -> Optional[User]:
        """Find user by email"""
        for user in self._users.values():
            if user.email == email:
                return user
        return None
    
    def __len__(self) -> int:
        return len(self._users)
    
    def __iter__(self):
        return iter(self._users.values())

